### PR TITLE
Sequence/Recorder updates

### PR DESCRIPTION
**Pull request by openbots-ff**
- Fixed bug when opening/closing sequences by creating a separate sequence form instead of creating a new instance of ScriptBuilder.
- Disabled the ability to drag sequences into sequences, and removed SequenceCommand from the command list on the left hand side in the Sequence form. 
- Varaibles/Arguments/Elements can be instantiated in a sequence, and if the sequence is saved, the parent studio will be updated accordingly
- Fixed element cloning bug in Advanced UI Recorder. 
- Added the ability to extract the name of a window when clicked. This allows the Advanced UI Recorder to produce an ActivateWindowCommand when a user clicks on a different application.
- Using the above concept, a Capture Window Name helper button was added to input/window commands to populate the window name input by clicking on the desired window.

### CODE REVIEW

File: OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderProject.cs [hunk 1]
```diff
@@ -187,6 +192,71 @@ public DialogResult AddProject()
...
+                uiScriptTabControl.TabPages.Clear();
+
+                //Open Main
+                OpenFile(mainFilePath);
+                //show success dialog
+                Notify("Project has been opened successfully!", Color.White);
+            }
+            catch (Exception ex)
+            {
+                //show fail dialog
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderProject.cs:230]: Not sure which convention you're using for comments but some start with lowercase and others like on line 223 are uppercase

File: OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderProject.cs [hunk 2]
```diff
@@ -187,6 +192,71 @@ public DialogResult AddProject()
...
+                AContainer = _builder.Build();
+
+                _mainFileName = mainFileName;
+                ScriptProject = project;
+                ScriptProjectPath = projectPath;
+                uiScriptTabControl.TabPages.Clear();
+
+                //Open Main
+                OpenFile(mainFilePath);
+                //show success dialog
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderProject.cs:225]: Not sure which convention you're using for comments but some start with lowercase and others like on line 223 are uppercase

File: OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderUIButtons.cs [hunk 3]
```diff
@@ -724,8 +725,9 @@ private void OpenSettingsManager()
             newSettings.ShowDialog();
 
             //reload app settings
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderUIButtons.cs:727]: See other comments about naming conventions for comments

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+            OpenVariableManager();
+        }
+
+        private void OpenVariableManager()
+        {
+            frmScriptVariables scriptVariableEditor = new frmScriptVariables
+            {
+                ScriptName = uiScriptTabControl.SelectedTab.Name,
+                ScriptVariables = ScriptVariables,
+                ScriptArguments = SriptArguments
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs:42]: second ScriptArguments has a typo

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        {
+            OpenArgumentManager();
+        }
+
+        private void OpenArgumentManager()
+        {
+            frmScriptArguments scriptArgumentEditor = new frmScriptArguments
+            {
+                ScriptName = uiScriptTabControl.SelectedTab.Name,
+                ScriptArguments = SriptArguments,
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs:67]: See previous typo comment

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+                ScriptVariables = ScriptVariables
+            };
+
+            if (scriptArgumentEditor.ShowDialog() == DialogResult.OK)
+            {
+                SriptArguments = scriptArgumentEditor.ScriptArguments;
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs:73]: First ScriptArguments has a typo

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+                if (!uiScriptTabControl.SelectedTab.Text.Contains(" *"))
+                    uiScriptTabControl.SelectedTab.Text += " *";
+
+                dgvArguments.DataSource = new BindingList<ScriptArgument>(SriptArguments);
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs:77]: ScriptArguments typo

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+                }
+
+                //end loop was found first
+                if (beginLoopValidationCount < 0)
+                {
+                    Notify("Please verify the ordering of your loops.", Color.Yellow);
+                    return;
+                }
+
+                //end if was found first
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs:224]: See comment about comment naming conventions

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+                else if (item.Tag.GetType().Name == "BeginSwitchCommand")
+                {
+                    beginSwitchValidationCount++;
+                }
+                else if (item.Tag.GetType().Name == "EndSwitchCommand")
+                {
+                    beginSwitchValidationCount--;
+                }
+
+                //end loop was found first
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs:217]: See comment about comment naming conventions

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+                }
+
+                if (beginSwitchValidationCount < 0)
+                {
+                    Notify("Please verify the ordering of your switch/case blocks.", Color.Yellow);
+                    return;
+                }
+            }
+
+            //extras were found
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs:250]: See comment about comment naming conventions

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceVariableArgument.cs [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+        private void dgvArguments_CellValueChanged(object sender, DataGridViewCellEventArgs e)
+        {
+            try
+            {
+                if (e.RowIndex != -1)
+                {
+                    DataGridView dgv = (DataGridView)sender;
+                    DataGridViewComboBoxCell cb = (DataGridViewComboBoxCell)dgv.Rows[e.RowIndex].Cells[2];
+
+                    //Sets value cell to read only if the argument direction is set to Out
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceVariableArgument.cs:178]: See comment about comment naming conventions

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceVariableArgument.cs [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+            }                       
+        }
+
+        private void dgvArguments_CurrentCellDirtyStateChanged(object sender, EventArgs e)
+        {
+            try
+            {
+                DataGridView dgv = (DataGridView)sender;
+
+                //This fires the cell value changed handler above
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceVariableArgument.cs:202]: See comment about comment naming conventions

File: OpenBots.Studio/UI/Forms/Supplement Forms/frmAdvancedUIElementRecorder.cs [hunk 13]
```diff
@@ -160,13 +165,30 @@ private void GlobalHook_MouseEvent(object sender, MouseCoordinateEventArgs e)
...
                 try
-                {
+                {                    
                     if (_isRecording)
                     {
+                        DPoint point = new DPoint((int)e.MouseCoordinates.X, (int)e.MouseCoordinates.Y);
+                        var window = User32Functions.WindowFromPoint(point);
+
+                        User32Functions.GetWindowThreadProcessId(window, out uint processId);
+                        Process process = Process.GetProcessById((int)processId);
```
Comment (openbots-nc) [inline OpenBots.Studio/UI/Forms/Supplement Forms/frmAdvancedUIElementRecorder.cs:175]: Server now uses the term "automation."  May want to change to keep consistent?

Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/Supplement Forms/frmAdvancedUIElementRecorder.cs:175]: This isn't a server process. This is from System.Diagnostics

File: OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderProject.cs (hunk 1 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderProject.cs:230]: resolved in 63f38b56

File: OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderProject.cs (hunk 2 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderProject.cs:230]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/63f38b56b74163c6219f8ae90bef5832933b6332

File: OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderUIButtons.cs (hunk 3 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/ScriptBuilder Forms/frmScriptBuilderUIButtons.cs:728]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/63f38b56b74163c6219f8ae90bef5832933b6332

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs (hunk 8 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs:224]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/63f38b56b74163c6219f8ae90bef5832933b6332

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs (hunk 9 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs:217]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/63f38b56b74163c6219f8ae90bef5832933b6332

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs (hunk 10 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs:250]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/63f38b56b74163c6219f8ae90bef5832933b6332

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceVariableArgument.cs (hunk 11 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceVariableArgument.cs]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/63f38b56b74163c6219f8ae90bef5832933b6332

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceVariableArgument.cs (hunk 12 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceVariableArgument.cs]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/63f38b56b74163c6219f8ae90bef5832933b6332

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs (hunk 4 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/4e142d77150428b358ce30dfdedcbdd7236ef478

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs (hunk 5 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/4e142d77150428b358ce30dfdedcbdd7236ef478

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs (hunk 6 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/4e142d77150428b358ce30dfdedcbdd7236ef478

File: OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs (hunk 7 above)
Comment (openbots-ff) [reply OpenBots.Studio/UI/Forms/Sequence Forms/frmSequenceUIButtons.cs]: resolved in https://github.com/OpenBotsAI/OpenBots.Studio/pull/95/commits/4e142d77150428b358ce30dfdedcbdd7236ef478

**Review by openbots-ch — APPROVED** [review]

**Review by openbots-nc — APPROVED** [review]

**Pull request merged**